File: secondMiniTime.py
from collections import defaultdict
import heapq


class Solution:
    def secondMinimum(self, n, edges, time, change):
        adj = defaultdict(list)
        for edge in edges:
            a, b = edge[0], edge[1]
            adj[a].append(b)
            adj[b].append(a)

        dist1 = [float('inf')] * (n + 1)
        dist2 = [float('inf')] * (n + 1)
        freq = [0] * (n + 1)

        pq = []
        heapq.heappush(pq, [1, 0])
        dist1[1] = 0

        while pq:
            node, time_taken = heapq.heappop(pq)

            freq[node] += 1

            if freq[node] == 2 and node == n:
                return time_taken

            if (time_taken // change) % 2 == 1:
                time_taken = change * ((time_taken // change) + 1) + time
            else:
                time_taken += time

            if node not in adj:
                continue

            for neighbor in adj[node]:
                if freq[neighbor] == 2:
                    continue

                if dist1[neighbor] > time_taken:
                    dist2[neighbor] = dist1[neighbor]
                    dist1[neighbor] = time_taken
                    heapq.heappush(pq, [neighbor, time_taken])
                elif dist2[neighbor] > time_taken and dist1[neighbor] != time_taken:
                    dist2[neighbor] = time_taken
                    heapq.heappush(pq, [neighbor, time_taken])

        return 0

File: secondMiniTime.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long 

 class Solution{
    
     public:

         int secondMinimumTime(int n, vector<int>&edges, int time, int change){
              vector<vector<int>>adj(n+1);

              for(auto edge: edges){
                  adj[edge[0]].push_back(edge[1]);
                  adj[edge[1]].push_back(edge[0]);
              }

              priority_queue<pair<int,int>, vector<pair<int,int>>, greater<int,int>>>minHeap;

              vector<int>dist1(n+1, INT_MAX), dist2(n+1,INT_MAX), freq(n+1,0);

              minHeap.push({0,1});

              while(!minHeap.empty()){
                   [timeTaken, node] = minHeap.top();
                   minHeap.pop();

                   freq[node]++;

                   if(freq[node] == 2 and node == n){
                       return timeTaken;
                   }

                   if((timeTaken/change)%2){
                       timeTaken = change * (timeTaken/change + 1) + time;
                   }else{
                      timeTaken = timeTaken + time;
                   }

                   for(auto neighbor: adj[node]){

                        if(dist1[neighbor] > timeTaken){
                            dist2[neighbor] = dist1[neighbor];
                            dist1[neighbor] = timeTaken;
                            minHeap.push({timeTaken, neighbor});
                        }else if(dist2[neighbor] > timeTaken and dist1[neighbor] != timeTaken){
                            dist2[neighbor] = timeTaken;
                            minHeap.push({timeTaken,neighbor});
                        }
                   }

              }

              return 0;
         }
 };     
